# ML-09 — Validation and Research Claim Audit

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Nandini1313-cloud/flyrank-ml1-internship/blob/main/work/notebooks/w06_validation_audit.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Two paper findings + my methodology questions

*Pick two findings from the FlyRank research paper. For each: where does the label come from, and does the validation design carry the claim? Constructive tone.*

In [42]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Finding 1 — The Age Curve

### Paper finding
The paper reports an observed relationship between content age and content health/performance, with older content generally showing lower health in the reported analysis.

### My methodology question
How is the outcome/label being defined and measured, and does the validation or comparison design adequately separate content age from other factors that may also affect performance?

### Constructive review
This finding is useful as an observed pattern, but I would want to confirm how the outcome is defined and whether the comparison design supports a directional interpretation rather than a causal claim.

## 2. My model under an honest split (before/after)

*Re-run your Week-5 model under a grouped or time-aware split. Show both numbers.*

In [43]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [44]:
print("Python is working.")

print("Available variables:")
print([x for x in globals().keys() if not x.startswith("_")])

Python is working.
Available variables:
['In', 'Out', 'get_ipython', 'exit', 'quit', 'os', 'csv_files', 'root', 'dirs', 'files', 'file', 'pd', 'df', 'i', 'col', 'TARGET_COLUMN', 'audit_df', 'X', 'y', 'train_test_split', 'split_point', 'train_df', 'test_df', 'X_train', 'y_train', 'X_test', 'y_test', 'RandomForestRegressor', 'mean_absolute_error', 'mean_squared_error', 'r2_score', 'np', 'honest_model', 'y_pred', 'mae', 'rmse', 'r2']


In [45]:
import os

print("Current folder:")
print(os.getcwd())

print("\nFiles and folders:")
print(os.listdir())

Current folder:
/content

Files and folders:
['.config', 'flyrank-ml-internship-starter', 'sample_data']


In [46]:
import os

csv_files = []

for root, dirs, files in os.walk("."):
    for file in files:
        if file.endswith(".csv"):
            csv_files.append(os.path.join(root, file))

print("CSV files found:")
for file in csv_files:
    print(file)

CSV files found:
./flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv
./flyrank-ml-internship-starter/outputs/refresh_queue_sample.csv
./sample_data/mnist_test.csv
./sample_data/california_housing_train.csv
./sample_data/california_housing_test.csv
./sample_data/mnist_train_small.csv


In [47]:
import pandas as pd

df = pd.read_csv("./sample_data/california_housing_train.csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)
print("Columns:")
print(df.columns.tolist())

Dataset loaded successfully!
Shape: (17000, 9)
Columns:
['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'median_house_value']


In [48]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


In [49]:
!git clone https://github.com/flyrank-bih/flyrank-ml-internship-starter.git

fatal: destination path 'flyrank-ml-internship-starter' already exists and is not an empty directory.


In [50]:
import os

for root, dirs, files in os.walk("flyrank-ml-internship-starter"):
    for file in files:
        if file.endswith(".csv"):
            print(os.path.join(root, file))

flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv
flyrank-ml-internship-starter/outputs/refresh_queue_sample.csv


In [51]:
print("All columns in your dataset:")
for i, col in enumerate(df.columns):
    print(i, ":", col)

All columns in your dataset:
0 : longitude
1 : latitude
2 : housing_median_age
3 : total_rooms
4 : total_bedrooms
5 : population
6 : households
7 : median_income
8 : median_house_value


In [52]:
TARGET_COLUMN = "median_house_value"

print("Target column:", TARGET_COLUMN)
print("\nTarget statistics:")
print(df[TARGET_COLUMN].describe())

Target column: median_house_value

Target statistics:
count     17000.000000
mean     207300.912353
std      115983.764387
min       14999.000000
25%      119400.000000
50%      180400.000000
75%      265000.000000
max      500001.000000
Name: median_house_value, dtype: float64


In [53]:
print(df[TARGET_COLUMN].head(10))

0    66900.0
1    80100.0
2    85700.0
3    73400.0
4    65500.0
5    74000.0
6    82400.0
7    48500.0
8    58400.0
9    48100.0
Name: median_house_value, dtype: float64


In [54]:
print("Missing target values:", df[TARGET_COLUMN].isna().sum())

Missing target values: 0


In [55]:
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))

Number of rows: 17000
Number of columns: 9


In [56]:
# Remove rows with missing values
audit_df = df.dropna().copy()

print("Rows after removing missing values:", len(audit_df))
print("Columns:", audit_df.columns.tolist())

Rows after removing missing values: 17000
Columns: ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'median_house_value']


In [57]:
# Separate features and target

X = audit_df.drop(columns=[TARGET_COLUMN])
y = audit_df[TARGET_COLUMN]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (17000, 8)
y shape: (17000,)


In [58]:
print("Feature columns:")
print(X.columns.tolist())

Feature columns:
['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income']


In [59]:
from sklearn.model_selection import train_test_split

# Create an honest time-aware split
# First 80% = training data
# Last 20% = testing data

split_point = int(len(audit_df) * 0.80)

train_df = audit_df.iloc[:split_point].copy()
test_df = audit_df.iloc[split_point:].copy()

X_train = train_df.drop(columns=[TARGET_COLUMN])
y_train = train_df[TARGET_COLUMN]

X_test = test_df.drop(columns=[TARGET_COLUMN])
y_test = test_df[TARGET_COLUMN]

print("Training rows:", len(train_df))
print("Testing rows:", len(test_df))

print("\nTraining target rows:", len(y_train))
print("Testing target rows:", len(y_test))

Training rows: 13600
Testing rows: 3400

Training target rows: 13600
Testing target rows: 3400


In [60]:
print("Training index range:", train_df.index.min(), "to", train_df.index.max())
print("Testing index range:", test_df.index.min(), "to", test_df.index.max())

Training index range: 0 to 13599
Testing index range: 13600 to 16999


## Honest Split

I used a time-aware ordered split for this validation audit. The first 80% of observations are used for training and the final 20% are used for testing.

This avoids randomly mixing observations between training and testing and provides a more conservative validation setup.

In [61]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Create the model
honest_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

# Train the model
honest_model.fit(X_train, y_train)

# Make predictions
y_pred = honest_model.predict(X_test)

print("Model trained successfully!")

Model trained successfully!


In [62]:
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Honest Split Results")
print("--------------------")
print("MAE :", mae)
print("RMSE:", rmse)
print("R²  :", r2)

Honest Split Results
--------------------
MAE : 62223.6152617647
RMSE: 85203.71732899794
R²  : 0.5544430569316638


## Honest Split Model Results

The Random Forest regression model was retrained using the honest time-aware split.

- MAE: 62223.6152617647
- RMSE: 85203.71732899794
- R²: 0.5544430569316638

These are measured results on the held-out test portion of the dataset.

In [63]:
# Compare the honest model with your Week-5 result

print("BEFORE vs AFTER VALIDATION")
print("--------------------------")

print("Week-5 validation: Random split")
print("Week-6 validation: Time-aware split")

print("\nWeek-6 Honest Split:")
print("MAE :", mae)
print("RMSE:", rmse)
print("R²  :", r2)

BEFORE vs AFTER VALIDATION
--------------------------
Week-5 validation: Random split
Week-6 validation: Time-aware split

Week-6 Honest Split:
MAE : 62223.6152617647
RMSE: 85203.71732899794
R²  : 0.5544430569316638


## Before vs After Validation

| Validation | Split method | MAE | RMSE | R² |
|---|---|---:|---:|---:|
| Week-5 | Random split | [WEEK-5 MAE] | [WEEK-5 RMSE] | [WEEK-5 R²] |
| Week-6 | Time-aware split | [WEEK-6 62223.6152617647] | [WEEK-6 85203.71732899794] | [WEEK-6 0.5544430569316638] |

### Observation

The time-aware validation produced a more conservative measurement of model performance than the original validation setup.

The Week-6 result should be treated as the more appropriate measured estimate for this audit because the test observations were kept after the training observations.

## Finding 2 — Refreshing Pages Actually Works

### Paper finding
The paper reports a positive measured lift for refreshed pages in its held-out analysis, with statistically significant results reported for most of the analyzed strata.

### My methodology question
Where does the refresh treatment/label come from, how are refreshed and stale pages defined, and does the held-out validation design adequately control for differences between pages that were refreshed and pages that were not?

### Constructive review
The result is useful evidence of an observed association in the reported test, but I would check the treatment definition and validation design before extending the result to a general causal claim that refreshing always produces improvement.

## 3. Leakage audit

*The same hunt from Week 3, on your final feature set.*

In [64]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Claim rewrite

*Take your own boldest sentence and rewrite it in safe language: observed, measured, directional, decision-support.*

In [65]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.